lib/agent.js

```diff
@@ -0,0 +1,215 @@
+    ? new Client(origin, opts)
+    : new Pool(origin, opts)
+}
+
+const WeakRef = global.WeakRef || class DummyWeakRef {
+  constructor (value) {
+    this.value = value
+  }
+
+  deref () {
+    return this.value
+  }
+}
+const FinalizationRegistry = global.FinalizationRegistry || class DummyFinalizer {
+  register () {}
+}
+
+class Agent extends Dispatcher {
+  constructor ({ factory = defaultFactory, maxRedirections = 0, ...options } = {}) {
+    super()
+
+    if (typeof factory !== 'function') {
+      throw new InvalidArgumentError('factory must be a function.')
+    }
+
+    if (!Number.isInteger(maxRedirections) || maxRedirections < 0) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    this[kOptions] = JSON.parse(JSON.stringify(options))
+    this[kMaxRedirections] = maxRedirections
+    this[kFactory] = factory
+    this[kClients] = new Map()
+    this[kFinalizer] = new FinalizationRegistry(key => {
+      const ref = this[kClients].get(key)
+      if (ref !== undefined && ref.deref() === undefined) {
+        this[kClients].delete(key)
+      }
+    })
+    this[kClosed] = false
+    this[kDestroyed] = false
+
+    const agent = this
+
+    this[kOnDrain] = function onDrain (origin, targets) {
+      agent.emit('drain', origin, [agent, ...targets])
+    }
+
+    this[kOnConnect] = function onConnect (origin, targets) {
+      agent.emit('connect', origin, [agent, ...targets])
+    }
+
+    this[kOnDisconnect] = function onDisconnect (origin, targets, err) {
+      agent.emit('disconnect', origin, [agent, ...targets], err)
+    }
+  }
+
+  get connected () {
+    let ret = 0
+    for (const { connected } of this[kClients].values()) {
+      ret += connected
+    }
+    return ret
+  }
+
+  get size () {
+    let ret = 0
+    for (const { size } of this[kClients].values()) {
+      ret += size
+    }
+    return ret
+  }
+
+  get pending () {
+    let ret = 0
+    for (const { pending } of this[kClients].values()) {
+      ret += pending
+    }
+    return ret
+  }
+
+  get running () {
+    let ret = 0
+    for (const { running } of this[kClients].values()) {
+      ret += running
+    }
+    return ret
+  }
+
+  dispatch (opts, handler) {
+    if (!handler || typeof handler !== 'object') {
+      throw new InvalidArgumentError('handler')
+    }
+
+    try {
+      if (!opts || typeof opts !== 'object') {
+        throw new InvalidArgumentError('opts must be a object.')
+      }
+
+      if (typeof opts.origin !== 'string' || opts.origin === '') {
+        throw new InvalidArgumentError('opts.origin must be a non-empty string.')
+      }
+
+      if (this[kDestroyed]) {
+        throw new ClientDestroyedError()
+      }
+
+      if (this[kClosed]) {
+        throw new ClientClosedError()
+      }
+
+      const ref = this[kClients].get(opts.origin)
+
+      let client = ref ? ref.deref() : null
+      if (!client) {
+        client = this[kFactory](opts.origin, this[kOptions])
```

Comment:
is this a pool?

Reply:
Can be.

Reply:
Can we call the variable dispatcher? so it's clear it's not a client.

Reply:
Done